plugins/outputs/postgresql/postgresql.go

```diff
@@ -0,0 +1,424 @@
+		poolConfig.MaxConns = 1
+	}
+
+	if p.LogLevel != "" {
+		poolConfig.ConnConfig.Logger = utils.PGXLogger{p.Logger}
+		poolConfig.ConnConfig.LogLevel, err = pgx.LogLevelFromString(p.LogLevel)
+		if err != nil {
+			return fmt.Errorf("invalid log level")
+		}
+	}
+
+	// Yes, we're not supposed to store the context. However since we don't receive a context, we have to.
+	p.dbContext, p.dbContextCancel = context.WithCancel(context.Background())
+	p.db, err = pgxpool.ConnectConfig(p.dbContext, poolConfig)
+	if err != nil {
+		p.Logger.Errorf("Couldn't connect to server\n%v", err)
+		return err
+	}
+	p.tableManager = NewTableManager(p)
+
+	if p.TagsAsForeignKeys {
+		p.tagsCache = freecache.NewCache(5 * 1024 * 1024) // 5MB
+	}
+
+	maxConns := int(p.db.Stat().MaxConns())
+	if maxConns > 1 {
+		p.writeChan = make(chan *TableSource)
+		p.writeWaitGroup = utils.NewWaitGroup()
+		for i := 0; i < maxConns; i++ {
+			p.writeWaitGroup.Add(1)
+			go p.writeWorker(p.dbContext)
+		}
+	}
+
+	return nil
+}
+
+// Close closes the connection(s) to the database.
+func (p *Postgresql) Close() error {
+	if p.writeChan != nil {
+		// We're using async mode. Gracefully close with timeout.
+		close(p.writeChan)
+		select {
+		case <-p.writeWaitGroup.C():
+		case <-time.NewTimer(time.Second * 5).C:
+		}
+	}
+
+	// Die!
+	p.dbContextCancel()
+	p.db.Close()
+	p.tableManager = nil
+	return nil
+}
+
+func (p *Postgresql) Write(metrics []telegraf.Metric) error {
+	tableSources := NewTableSources(p, metrics)
+
+	if p.db.Stat().MaxConns() > 1 {
+		return p.writeConcurrent(tableSources)
+	} else {
+		return p.writeSequential(tableSources)
+	}
+}
+
+func (p *Postgresql) writeSequential(tableSources map[string]*TableSource) error {
+	tx, err := p.db.Begin(p.dbContext)
+	if err != nil {
+		return fmt.Errorf("starting transaction: %w", err)
+	}
+	defer tx.Rollback(p.dbContext)
+
+	for _, tableSource := range tableSources {
+		err := p.writeMetricsFromMeasure(p.dbContext, tx, tableSource)
+		if err != nil {
+			if isTempError(err) {
+				return err
+			}
+			p.Logger.Errorf("write error (permanent, dropping sub-batch): %v", err)
+		}
+	}
+
+	if err := tx.Commit(p.dbContext); err != nil {
+		return fmt.Errorf("committing transaction: %w", err)
+	}
+	return nil
+}
+
+func (p *Postgresql) writeConcurrent(tableSources map[string]*TableSource) error {
+	for _, tableSource := range tableSources {
+		select {
+		case p.writeChan <- tableSource:
+		case <-p.dbContext.Done():
+			return nil
+		}
+	}
+	return nil
+}
+
+func (p *Postgresql) writeWorker(ctx context.Context) {
+	defer p.writeWaitGroup.Done()
+	for {
+		select {
+		case tableSource, ok := <-p.writeChan:
+			if !ok {
+				return
+			}
+			if err := p.writeRetry(ctx, tableSource); err != nil {
+				p.Logger.Errorf("write error (permanent, dropping sub-batch): %v", err)
+			}
+		case <-p.dbContext.Done():
+			return
+		}
+	}
+}
+
+// This is a subset of net.Error
+type maybeTempError interface {
+	error
+	Temporary() bool
+}
```

Comment:
Do we really need this? We can still call `isTempError` without defining the interface...